storage/range.go

```diff
@@ -281,42 +280,72 @@ func (r *Range) IsFirstRange() bool {
+	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
 }
 
-func (r *Range) setLease(l *proto.Lease) {
-	atomic.StorePointer(&r.lease, unsafe.Pointer(l))
+// newNotLeaderError returns a NotLeaderError intialized with the
+// replica for the last known holder of the leader lease.
+func (r *Range) newNotLeaderError() error {
+	err := &proto.NotLeaderError{}
+	if l := r.getLease(); l.RaftNodeID != 0 {
+		_, storeID := DecodeRaftNodeID(multiraft.NodeID(l.RaftNodeID))
+		_, err.Leader = r.Desc().FindReplica(storeID)
+	}
+	return err
 }
 
-func (r *Range) getLease() *proto.Lease {
-	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
+// requestLeaderLease sends a request to obtain or extend a leader lease for
+// this replica.
+func (r *Range) requestLeaderLease(term uint64) {
+	// Register a stopper task while lease request is in flight.
+	if !r.rm.Stopper().StartTask() {
+		return
+	}
+	defer r.rm.Stopper().FinishTask()
+	// Prepare a Raft command to get a leader lease for the replica
+	// of that group that lives in our store.
+	wallTime := r.rm.Clock().PhysicalNow()
+	// TODO: get duration from configuration, either as a config flag
+	// or, later, dynamically adjusted.
+	duration := int64(defaultLeaderLeaseDuration)
+	args := &proto.InternalLeaderLeaseRequest{
+		RequestHeader: proto.RequestHeader{Key: r.Desc().StartKey},
+		Lease: proto.Lease{
+			Expiration: wallTime + duration,
+			Duration:   duration,
+			Term:       term,
+			RaftNodeID: uint64(r.rm.RaftNodeID()),
+		},
+	}
+	r.AddCmd(proto.InternalLeaderLease, args, &proto.InternalLeaderLeaseResponse{}, true)
 }
 
-// canServiceCmd returns an error in the event that the range replica
-// cannot service the command as specified. This is of the case in
-// the event that the replica is not the leader.
-func (r *Range) canServiceCmd(method string, args proto.Request) error {
-	header := args.Header()
-	if !r.IsLeader() {
-		if !proto.IsReadOnly(method) || header.ReadConsistency == proto.CONSISTENT {
-			// TODO(spencer): when we happen to know the leader, fill it in here via replica.
-			return &proto.NotLeaderError{}
-		}
-	}
-	if proto.IsReadOnly(method) {
-		if header.ReadConsistency == proto.CONSENSUS {
-			return util.Errorf("consensus reads not implemented")
-		} else if header.ReadConsistency == proto.INCONSISTENT && header.Txn != nil {
-			return util.Errorf("cannot allow inconsistent reads within a transaction")
+// HasLeaderLease returns whether the leader lease is held by this
+// replica and the term in which it was held. It's possible that the
+// replica is not in fact the leader, but holding the lease means that
+// no other replica may be the leader either. Leases may not overlap,
+// though a gap between successive lease holders is expected.
+func (r *Range) HasLeaderLease() (bool, uint64) {
+	if l := r.getLease(); l.RaftNodeID != 0 {
+		if l.RaftNodeID == uint64(r.rm.RaftNodeID()) &&
+			r.rm.Clock().PhysicalNow() < l.Expiration {
+			return true, l.Term
```

Comment:
Once we make sure our leader lease requests are only ever proposed by the local node, it seems the Term isn't relevant for verifying the lease any more (term changes trigger some updates of course, but those are maybe better triggered from the store when an election event is received). If it remains relevant somehow, then this validation should happen in here and not at the callsite.